torch/csrc/deploy/interpreter/interpreter_impl.cpp

```diff
@@ -0,0 +1,463 @@
+      const std::shared_ptr<caffe2::serialize::PyTorchStreamReader>&
+          container_file_) override {
+    InitLockAcquire guard(interp_->init_lock_);
+    return wrap(interp_->get_package(container_file_));
+  }
+
+  PickledObject pickle(PythonObject container, PythonObject obj) override {
+    py::tuple result = interp_->save_storage(unwrap(container), unwrap(obj));
+    py::bytes bytes = py::cast<py::bytes>(result[0]);
+    py::list storages = py::cast<py::list>(result[1]);
+    py::list dtypes = py::cast<py::list>(result[2]);
+    auto container_file =
+        py::cast<std::shared_ptr<caffe2::serialize::PyTorchStreamReader>>(
+            result[3]);
+
+    std::vector<at::Storage> storages_c;
+    std::vector<at::ScalarType> dtypes_c;
+    for (size_t i = 0, N = storages.size(); i < N; ++i) {
+      storages_c.push_back(torch::createStorage(storages[i].ptr()));
+      dtypes_c.push_back(
+          reinterpret_cast<THPDtype*>(dtypes[i].ptr())->scalar_type);
+    }
+    return PickledObject{
+        bytes,
+        std::move(storages_c),
+        std::move(dtypes_c),
+        std::move(container_file)};
+  }
+  PythonObject unpickle_or_get(int64_t id, const PickledObject& obj) override {
+    py::dict objects = interp_->objects;
+    py::object id_p = py::cast(id);
+    if (objects.contains(id_p)) {
+      return wrap(objects[id_p]);
+    }
+
+    InitLockAcquire guard(interp_->init_lock_);
+    // re-check if something else loaded this before we acquired the
+    // init_lock_
+    if (objects.contains(id_p)) {
+      return wrap(objects[id_p]);
+    }
+
+    py::tuple storages(obj.storages_.size());
+    for (size_t i = 0, N = obj.storages_.size(); i < N; ++i) {
+      py::object new_storage =
+          py::reinterpret_steal<py::object>(torch::createPyObject(
+              obj.storages_[i], scalarTypeToTypeMeta(obj.types_[i])));
+      storages[i] = std::move(new_storage);
+    }
+    py::object result = interp_->load_storage(
+        id, obj.container_file_, py::bytes(obj.data_), storages);
+    return wrap(result);
+  }
+  void unload(int64_t id) override {
+    py::dict objects = interp_->objects;
+    py::object id_p = py::cast(id);
+    if (objects.contains(id_p)) {
+      objects.attr("__delitem__")(id_p);
+    }
+  }
+
+  IValue toIValue(PythonObject obj) const override {
+    return torch::jit::toTypeInferredIValue(unwrap(obj));
+  }
+
+  PythonObject call(PythonObject obj, at::ArrayRef<PythonObject> args)
+      override {
+    py::tuple m_args(args.size());
+    for (size_t i = 0, N = args.size(); i != N; ++i) {
+      m_args[i] = unwrap(args[i]);
+    }
+    return wrap(call(unwrap(obj), m_args));
+  }
+
+  PythonObject call(PythonObject obj, at::ArrayRef<IValue> args) override {
+    py::tuple m_args(args.size());
+    for (size_t i = 0, N = args.size(); i != N; ++i) {
+      m_args[i] = torch::jit::toPyObject(args[i]);
+    }
+    return wrap(call(unwrap(obj), m_args));
```

Comment:
is wrap/unwrap used to work around some python overhead so the common case uses existing python object references instead of looking up by string again?  i first thought maybe it was like a 'unique_ptr' for py::objects but i don't think that's it.

Reply:
wrap turns a py::object into a PythonObject. PythonObjects are the handles people use outside of interpreter_impl.cpp where pybind11 and python APIs are not allowed because the libraries are hidden.